code/model1.py

```diff
@@ -1,9 +1,91 @@
+        # embedding lookup for 150 users each have 50 dimension representation
+        self.embedding_layer = nn.Embedding(150, 50)
+        # first hidden layer, linear layer with weights 200x500
+        # this should be the size of <sender+receiver representation>
+        self.h1_layer = nn.Linear(100, 500)
+        # ReLU activation used
+        self.relu = nn.ReLU()
+        # final linear layer that outputs the predicted email representation
+        self.email_layer = nn.Linear(500, 50)
+
+    def forward(self, s_id, r_id):
+        """
+        Input are 2 autgrad variables representing sender_id and receiver_id. These ids will be used to lookup the
+        embeddings. Does a forward pass and returns the predicted email representation
+        :param s_id:
+        :param r_id
+        :return: email representation
+        """
+        # extract the sender and receiver embedding
+        s_emb = self.embedding_layer(s_id)
+        r_emb = self.embedding_layer(r_id)
+        # simple concatenation of sender and receiver embedding
+        sr_emb = torch.cat((s_emb, r_emb), 1)
+        h1 = self.relu(self.h1_layer(sr_emb))
+        email_reps = self.email_layer(h1)
+        return email_reps
+
+    def get_average_rep(self, word_reps):
+        """
+        Assumes that word_reps is a numpy 2d array with every row as vector representation of word.
+        Calculates the mean across all rows to get average email representation
+        :param word_reps:
+        :return:
+        """
+        return np.mean(word_reps, axis=0)
+
     def train(self, emails, w2v):
-        pass
+        loss_criteria = nn.MSELoss()
+        optimizer = optim.RMSprop(self.parameters(), lr=0.001, alpha=0.6, momentum=0.6)
+
+        for epoch in range(self.epochs):
+            epoch_loss = 0.0
+            for i in range(len(emails)):
+                sender_id = utils.get_userid(emails[i, 0])
+                email_word_reps = w2v.get_sentence(emails[i, 2])
+                # if no word_rep was found for any of the words in the emails, ignore this case
+                if len(email_word_reps) == 0:
+                    continue
+                email_rep = self.get_average_rep(email_word_reps)
+                recv_list = emails[i, 1].split('|')
+                for recv in recv_list:
+                    optimizer.zero_grad()
+                    recv_id = utils.get_userid(recv)
+                    # if sender or receiver is not an enron email id, we ignore this data point
+                    if sender_id is None or recv_id is None:
+                        continue
+                    # do the forward pass
+                    pred_email_rep = self.forward(autograd.Variable(torch.LongTensor([sender_id])),
+                                                  autograd.Variable(torch.LongTensor([recv_id])))
+                    # compute the loss
+                    loss = loss_criteria(pred_email_rep, autograd.Variable(torch.from_numpy(email_rep)))
```

Comment:
Some of these function calls can be pulled out of this loop and put into the outer one for speedups